### ISSUE TITLE
[Linux] Apps are still protected after excluding them from VPN protection

### ISSUE THREAD

**Issue by Cimbali** [mozilla-mobile/mozilla-vpn-client#6371]
**Describe the bug**
Setting an app as “excluded” does not cause its traffic to bypass the VPN.

**VPN version:**
2.16.0 (current master).

**Affected Platforms:**
Linux. Qt 6.4.2.

**Prerequisites:**
Check an app in the “Excluded Apps” list. In this example I use Firefox.

**Steps to reproduce:**
Connect to VPN and check whether traffic is redirected from the selected App (e.g. checking apparent IP on the internet)

**Expected result:**
Traffic should not be tunnelled for excluded apps.

**Actual result:**
Traffic should is tunnelled for excluded apps.

**Additional notes:**

<details><summary>I’ve tried to find the source of the error so here is a log of debug info. You can jump to conclusions below.</summary>

1. In VPN logs, when starting mozillavpn, Firefox is correctly identified:
```
[20.03.2023 15:13:20.344] (AppTracker) Debug: Control group created: /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-c74d8d25cbf442091dabd872d75a72d.scope
[20.03.2023 15:13:20.345] (DBusService) Debug: tracking: /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-0c74d8d25cbf442091dabd872d75a72d.scope appId: /usr/share/applications/firefox.desktop PID: 0
```

```sh
> cat /proc/`ps -o pid= -C firefox | tr -d ' '`/cgroup
0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-0c74d8d25cbf442091dabd872d75a72d.scope
```

2. When restarting firefox with mozilla vpn already started, Firefox is still correctly identified:
```
[20.03.2023 15:44:36.057] (AppTracker) Debug: Control group created: /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-5b3c9c2ecfa5452e8d6a44272749f2aa.scope
[20.03.2023 15:44:36.083] (DBusService) Debug: tracking: /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-5b3c9c2ecfa5452e8d6a44272749f2aa.scope appId: /usr/share/applications/firefox.desktop PID: 0
```

```sh
> cat /proc/`ps -o pid= -C firefox | tr -d ' '`/cgroup
0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-5b3c9c2ecfa5452e8d6a44272749f2aa.scope
```

3. Firefox is reported as a disabled app with the same id

Config reported in log file:
```
vpnDisabledApps -> [/usr/share/applications/firefox.desktop]
```

4. There are no traces of attempts to un-tunnel the cgroup traffic

All of the WireguardUtils / NetfilterGo activity in the log, with ip addresses removed:
```sh
> sed -rn 's/(Adding exclusion route for|Adding route to|Isolating tunnel address) [0-9a-f.:/]*$/\1 <IP>/;s/(Adding peer|found) [A-Za-z0-9.=]+/\1 <PEER>/;/WireguardUtilsLinux/p' Desktop/mozillavpn-2023-3-20.txt | uniq -cf2
      1 [20.03.2023 15:13:20.331] (WireguardUtilsLinux) Debug: NetfilterGo: Creating netfilter tables
      1 [20.03.2023 15:13:20.332] (WireguardUtilsLinux) Info: Setup split tunneling with unified cgroups (v2)
      1 [20.03.2023 15:13:20.332] (WireguardUtilsLinux) Debug: WireguardUtilsLinux created.
      1 [20.03.2023 15:13:28.659] (WireguardUtilsLinux) Debug: NetfilterGo: Starting netfilter tables for moz0
      1 [20.03.2023 15:13:28.660] (WireguardUtilsLinux) Debug: NetfilterGo: Isolating tunnel address <IP>
      2 [20.03.2023 15:13:28.660] (WireguardUtilsLinux) Debug: Adding exclusion route for <IP>
      1 [20.03.2023 15:13:28.660] (WireguardUtilsLinux) Debug: Adding peer <PEER>
      1 [20.03.2023 15:13:28.660] (WireguardUtilsLinux) Debug: NetfilterGo: Marking inbound traffic from server
    203 [20.03.2023 15:13:28.663] (WireguardUtilsLinux) Debug: Adding route to <IP>
      2 [20.03.2023 15:13:28.934] (WireguardUtilsLinux) Debug: found <PEER> handshake 1679325208
```

So no messages which would be expected:
- `Marking traffic from cgroup` or `Unable to mark traffic: no fwmark`
- `Excluding traffic from .*`
- `Setting .* to firewall state .*`

Looking at the `appLaunched` function I can only guess `m_excludedApps` is not updated properly:

https://github.com/mozilla-mobile/mozilla-vpn-client/blob/1fa1f75e8e3d4e2e376c820794db4fee37d86e91/src/apps/vpn/platforms/linux/daemon/dbusservice.cpp#L168-L178

Ticking/unticking the app does not work as a work-around. The following messages appear:
```log
[20.03.2023 16:17:30.115] (AppPermission) Debug: Enabled -- /usr/share/applications/firefox.desktop  for VPN
[20.03.2023 16:17:32.239] (AppPermission) Debug: Disabled -- /usr/share/applications/firefox.desktop  for VPN
```

But nothing else happens.

Note that that function emits a signal:
https://github.com/mozilla-mobile/mozilla-vpn-client/blob/1fa1f75e8e3d4e2e376c820794db4fee37d86e91/src/apps/vpn/apppermission.cpp#L106-L120

But I can’t figure out who subscribes to it.
</details>

TL;DR is I can check apps on the list without having the “Protect all apps” checkbox.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/VPN-4388)

**Comment by Cimbali** [comment]
Finally found a way to make something happen, which is
1. to connect the VPN
2. to manually call the `firewallApp()` function on the DBUS service, e.g. with:
```sh
> dbus-send --system --print-reply --dest=org.mozilla.vpn.dbus / org.mozilla.vpn.dbus.firewallApp string:/usr/share/applications/firefox.desktop string:excluded
method return time=1679332386.321834 sender=:1.613 -> destination=:1.641 serial=21 reply_serial=2
   boolean true
```

The log then shows:
```log
[20.03.2023 16:56:54.972] (DBusService) Debug: Setting /usr/share/applications/firefox.desktop to firewall state excluded
[20.03.2023 16:56:54.972] (WireguardUtilsLinux) Info: Excluding traffic from /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-5b3c9c2ecfa5452e8d6a44272749f2aa.scope
[20.03.2023 16:56:54.972] (WireguardUtilsLinux) Debug: NetfilterGo: Marking traffic from cgroup /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-5b3c9c2ecfa5452e8d6a44272749f2aa.scope
```

Unfortunately Firefox then loses all connectivity but that may be unrelated (?)

Monitoring the dbus activation shows an empty `"vpnDisabledApps":[]` which means the `HopConnection` is not inheriting the vpn’s disabled apps properly.

**Comment by Cimbali** [comment]
Reverting to v2.13.1 displays the checkbox currently:

![image](https://user-images.githubusercontent.com/6126377/226432908-6276c6e9-113b-44be-a6f6-67bce2649d24.png)

Current master:

![image](https://user-images.githubusercontent.com/6126377/226433463-9945fee9-936d-4631-9530-6483f23a2586.png)

Re-upgrading from v2.13.1 after setting the checkbox prints the expected messages:
```
[20.03.2023 18:30:36.577] (DBusService) Debug: Setting /usr/share/applications/firefox.desktop to firewall state excluded
[20.03.2023 18:30:36.577] (WireguardUtilsLinux) Info: Excluding traffic from /user.slice/user-1000.slice/user@1000.service/app.slice/app-firefox-bbdfd291e09a40c294e4b5eb407cf686.scope
[20.03.2023 18:30:36.577] (WireguardUtilsLinux) Debug: NetfilterGo: Unable to mark traffic: no fwmark
```

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Naomi Kirby same question as in https://mozilla-hub.atlassian.net/browse/VPN-4404 ( https://mozilla-hub.atlassian.net/browse/VPN-4404|smart-link ).

**Comment by Cimbali** [comment]
This is a regression from v2.13 which displayed the checkbox correctly on the same distribution/library versions/etc.

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Actually, this looks like a cross-platform version of https://mozilla-hub.atlassian.net/browse/VPN-4644 ( https://mozilla-hub.atlassian.net/browse/VPN-4644|smart-link ), that Basti is looking into. Might have a common origin.

**Comment by data-sync-user** [comment]
➤ Adrienne Davenport commented:

Duplicate of an issue that is fixed
